src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -710,9 +712,29 @@ public static SocketError Poll(SafeCloseSocket handle, int microseconds, SelectM
 
         public static SocketError Select(IList checkRead, IList checkWrite, IList checkError, int microseconds)
         {
-            IntPtr[] readfileDescriptorSet = Socket.SocketListToFileDescriptorSet(checkRead);
-            IntPtr[] writefileDescriptorSet = Socket.SocketListToFileDescriptorSet(checkWrite);
-            IntPtr[] errfileDescriptorSet = Socket.SocketListToFileDescriptorSet(checkError);
+            bool ShouldStackAlloc(IList list, out Span<IntPtr> span)
+            {
+                const int StackThreshold = 80; // arbitrary limit to avoid too much space on stack
+                if (list == null || list.Count == 0)
+                {
+                    span = default;
+                    return false;
+                }
+                if (list.Count >= StackThreshold)
+                {
+                    span = new IntPtr[list.Count + 1];
+                    return false;
+                }
+                span = default;
+                return true;
+            }
+
+            var readfileDescriptorSet = ShouldStackAlloc(checkRead, out var tmp) ? stackalloc IntPtr[checkRead.Count + 1] : tmp;
+            Socket.SocketListToFileDescriptorSet(checkRead, readfileDescriptorSet);
+            var writefileDescriptorSet = ShouldStackAlloc(checkWrite, out tmp) ? stackalloc IntPtr[checkWrite.Count + 1] : tmp;
+            Socket.SocketListToFileDescriptorSet(checkWrite, writefileDescriptorSet);
+            var errfileDescriptorSet = ShouldStackAlloc(checkError, out tmp) ? stackalloc IntPtr[checkError.Count + 1] : tmp;
```

Comment:
Nit: all of these `var`s => `Span<IntPtr>`

Reply:
done